pkg/apiserver/handlers/webhook/mutation_labels.go

```diff
@@ -0,0 +1,218 @@
+	Value interface{} `json:"value,omitempty"`
+}
+
+func HandleMutationLabels() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		klog.V(2).Info("Received request to mutate resource labels")
+		var reqBody []byte
+		if r.Body != nil {
+			reqBody, _ = ioutil.ReadAll(r.Body)
+		}
+		if len(reqBody) == 0 {
+			klog.Errorf("Mutation webhook labelsmutator received empty request body")
+			http.Error(w, "empty request body", http.StatusBadRequest)
+			return
+		}
+		// verify the content type is accurate
+		contentType := r.Header.Get("Content-Type")
+		if contentType != "application/json" {
+			klog.Errorf("Invalid Content-Type=%s, expected application/json", contentType)
+			http.Error(w, "invalid Content-Type, expected `application/json`", http.StatusUnsupportedMediaType)
+			return
+		}
+		var admissionResponse *admv1.AdmissionResponse
+		ar := admv1.AdmissionReview{}
+		ar.TypeMeta.Kind = "AdmissionReview"
+		ar.TypeMeta.APIVersion = "admission.k8s.io/v1"
+		if err := json.Unmarshal(reqBody, &ar); err != nil {
+			klog.Errorf("Label mutation received incorrect body")
+			admissionResponse = getAdmissionResponseForErr(err)
+		} else {
+			admissionResponse = mutateResourceLabels(&ar)
+		}
+		aReview := admv1.AdmissionReview{}
+		aReview.TypeMeta.Kind = "AdmissionReview"
+		aReview.TypeMeta.APIVersion = "admission.k8s.io/v1"
+		if admissionResponse != nil {
+			aReview.Response = admissionResponse
+			if ar.Request != nil {
+				aReview.Response.UID = ar.Request.UID
+			}
+		}
+		resp, err := json.Marshal(aReview)
+		if err != nil {
+			klog.Errorf("Unable to encode response during mutation: %v", err)
+			http.Error(w, fmt.Sprintf("could not encode response: %v", err), http.StatusInternalServerError)
+		}
+		klog.V(2).Infof("Writing mutation response to MutationAdmissionHook")
+		if _, err := w.Write(resp); err != nil {
+			klog.Errorf("Unable to write response during mutation: %v", err)
+			http.Error(w, fmt.Sprintf("could not write response: %v", err), http.StatusInternalServerError)
+		}
+	}
+}
+
+// mutateResourceLabels mutates the resource labels and inserts Antrea required labels.
+func mutateResourceLabels(ar *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var result *metav1.Status
+	var patch []byte
+	var allowed bool
+	var msg string
+	patchType := admv1.PatchTypeJSONPatch
+
+	op := ar.Request.Operation
+	curRaw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	switch ar.Request.Kind.Kind {
+	case "Namespace":
+		klog.V(2).Info("Mutating Namespace labels")
+		var curNS, oldNS corev1.Namespace
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curNS); err != nil {
+				klog.Errorf("Error de-serializing current Namespace")
+				return getAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldNS); err != nil {
+				klog.Errorf("Error de-serializing old Namespace")
+				return getAdmissionResponseForErr(err)
+			}
+		}
```

Comment:
it seems that we don't actually need the old object, so maybe we can skip unmarshalling it altogether?